File: src/main/java/cz/ememsoft/minibank/mapper/CustomerMapper.java
package cz.ememsoft.minibank.mapper;

import cz.ememsoft.minibank.dto.CustomerDto;
import cz.ememsoft.minibank.entity.Customer;
import org.mapstruct.AfterMapping;
import org.mapstruct.BeanMapping;
import org.mapstruct.Mapper;
import org.mapstruct.MappingConstants;
import org.mapstruct.MappingTarget;
import org.mapstruct.NullValuePropertyMappingStrategy;
import org.mapstruct.ReportingPolicy;

@Mapper(unmappedTargetPolicy = ReportingPolicy.IGNORE, componentModel = MappingConstants.ComponentModel.SPRING)
public interface CustomerMapper {
    Customer toEntity(CustomerDto customerDto);

    @AfterMapping
    default void linkAccount(@MappingTarget Customer customer) {
        customer.getAccounts().forEach(account -> account.setCustomerId(customer));
    }

    CustomerDto toDto(Customer customer);

    @BeanMapping(nullValuePropertyMappingStrategy = NullValuePropertyMappingStrategy.IGNORE)
    Customer partialUpdate(CustomerDto customerDto, @MappingTarget Customer customer);
}

File: src/main/kotlin/cz/ememsoft/minibank/mapper/CustomerMapper.kt
package cz.ememsoft.minibank.mapper

import cz.ememsoft.minibank.api.dto.CustomerDto
import cz.ememsoft.minibank.api.dto.request.CustomerSaveRequest
import cz.ememsoft.minibank.entity.CustomerEntity
import org.mapstruct.Mapper
import org.mapstruct.MappingConstants

@Mapper(componentModel = MappingConstants.ComponentModel.SPRING)
interface CustomerMapper {

    // Map API request to internal DTO
    fun toDto(request: CustomerSaveRequest): CustomerDto

    // Map DTO to Entity
    fun toEntity(dto: CustomerDto): CustomerEntity

    // Map Entity back to DTO
    fun toDtoFromEntity(entity: CustomerEntity): CustomerDto
}
